phd-tests/framework/src/serial/vt80x24.rs

```diff
@@ -0,0 +1,174 @@
+    /// cells each.
+    surface: Surface,
+
+    /// The parsing state machine that converts incoming bytes into VT100
+    /// commands.
+    parser: Parser,
+}
+
+impl Vt80x24 {
+    pub(super) fn new() -> Self {
+        Self {
+            waiter: None,
+            parser: Parser::new(),
+            surface: Surface::new(80, 24),
+        }
+    }
+
+    /// Converts a list of VT100 actions into a set of changes to the 80x24
+    /// buffer, applies those changes, and attempts to satisfy the active wait.
+    fn apply_actions(&mut self, actions: &[Action]) {
+        // Unfortunately, there is no one-to-one mapping between VT100 actions
+        // and `Change`s to a termwiz `Surface`. The match below is enough to
+        // make buffering work for simple terminals that only print characters,
+        // position the cursor, and use the 'erase to end of line' VT100
+        // command. These commands are (fortunately) each representable in a
+        // single `Change`. More complex commands (e.g. 'erase to start of
+        // line') will have to be composed of multiple `Change`s if support for
+        // them is needed.
+        let to_change = |action: &Action| -> Option<Change> {
+            let change = match action {
+                Action::Print(c) => Some(Change::from(*c)),
+                Action::PrintString(s) => Some(Change::from(s)),
+                Action::Control(ctrl) => match ctrl {
+                    ControlCode::LineFeed => Some(Change::from('\n')),
+                    ControlCode::CarriageReturn => Some(Change::from('\r')),
+                    _ => None,
+                },
+                Action::CSI(csi) => match csi {
+                    CSI::Cursor(Cursor::Position { line, col }) => {
+                        Some(make_absolute_cursor_position(
+                            col.as_zero_based() as usize,
+                            line.as_zero_based() as usize,
+                        ))
+                    }
+                    CSI::Edit(Edit::EraseInLine(
+                        EraseInLine::EraseToEndOfLine,
+                    )) => {
+                        Some(Change::ClearToEndOfLine(ColorAttribute::Default))
+                    }
+                    _ => None,
+                },
+                _ => None,
+            };
+
+            trace!(?action, ?change, "termwiz VT100 action");
+            change
+        };
+
+        let changes = actions.iter().filter_map(to_change).collect();
+        let seq = self.surface.add_changes(changes);
+        self.surface.flush_changes_older_than(seq);
+
+        if let Some(waiter) = self.waiter.take() {
+            self.satisfy_or_set_wait(waiter);
+        }
+    }
+
+    /// Attempts to satisfy the wait described by `waiter`. If the wait is not
+    /// immediately satisfiable, stores `waiter` to try again later.
+    fn satisfy_or_set_wait(&mut self, waiter: OutputWaiter) {
+        assert!(self.waiter.is_none());
```

Comment:
maybe we should add an assertion like this:
```suggestion
        assert!(self.waiter.is_none());
        let too_long = waiter.wanted.lines().find(|line| line.len() > 80);
        assert_eq!(too_long, None, "a wait request that contains a line longer than 80 columns will never be satisfied");
```
 
i'd prefer my test to get a loud failure telling me what i did wrong rather than a wait that silently doesn't complete for unclear reasons...